presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -459,9 +503,189 @@ else if (!result.getBlocked().isDone()) {
+        if (newRoot != fragment.getRoot()) {
+            oldToNewFragmentMapping.put(fragment, new PlanFragment(
+                    fragment.getId(),
+                    newRoot,
+                    fragment.getVariables(),
+                    fragment.getPartitioning(),
+                    scheduleOrder(newRoot),
+                    fragment.getPartitioningScheme(),
+                    fragment.getStageExecutionDescriptor(),
+                    fragment.isOutputTableWriterFragment(),
+                    fragment.getStatsAndCosts(),
+                    fragment.getJsonRepresentation()));
+        }
+    }
+
+    /**
+     * Utility function that rebuild a StreamingPlanSection, re-create stageExecutionAndScheduler for each of its stage, and finally update the stageExecutions map.
+     */
+    private void updateStageExecutions(StreamingPlanSection section, Map<PlanFragment, PlanFragment> oldToNewFragment)
+    {
+        StreamingPlanSection newSection = new StreamingPlanSection(rewriteStreamingSubPlan(section.getPlan(), oldToNewFragment), section.getChildren());
+        PlanFragment sectionRootFragment = newSection.getPlan().getFragment();
+        Optional<int[]> bucketToPartition;
+        OutputBuffers outputBuffers;
+        ExchangeLocationsConsumer locationsConsumer;
+        if (isRootFragment(sectionRootFragment)) {
+            bucketToPartition = Optional.of(new int[1]);
+            outputBuffers = createInitialEmptyOutputBuffers(sectionRootFragment.getPartitioningScheme().getPartitioning().getHandle())
+                    .withBuffer(new OutputBufferId(0), BROADCAST_PARTITION_ID)
+                    .withNoMoreBufferIds();
+            OutputBufferId rootBufferId = getOnlyElement(outputBuffers.getBuffers().keySet());
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) ->
+                    updateQueryOutputLocations(queryStateMachine, rootBufferId, tasks, noMoreExchangeLocations);
+        }
+        else {
+            bucketToPartition = Optional.empty();
+            outputBuffers = createDiscardingOutputBuffers();
+            locationsConsumer = (fragmentId, tasks, noMoreExchangeLocations) -> {};
+        }
+        SectionExecution sectionExecution = sectionExecutionFactory.createSectionExecutions(
+                session,
+                newSection,
+                locationsConsumer,
+                bucketToPartition,
+                outputBuffers,
+                summarizeTaskInfo,
+                remoteTaskFactory,
+                splitSourceFactory,
+                0);
+        addStateChangeListeners(sectionExecution);
+        Map<StageId, StageExecutionAndScheduler> updatedStageExecutions = sectionExecution.getSectionStages().stream()
+                .collect(toImmutableMap(execution -> execution.getStageExecution().getStageExecutionId().getStageId(), identity()));
+        synchronized (this) {
+            stageExecutions.putAll(updatedStageExecutions);
```

Comment:
ditto: do we need to remove old stage executions from `stageExecutions` ?

Reply:
Samewise, stageExecutions is a map from StageId -> StageExecutionAndScheduler. The rewritten stages have the same id from the old one (a bit different from retry because we optimize them right before we schedule & execute them), and so we actually replace the old ones, not simply inserting a batch of new ones.